enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,993 @@
+present in `pkg/route/validation/validation_test.go`
+
+Please note that all values will be quoted before interpolating them into `haproxy.conf`.  Specifically, any single-quote present in the value is escaped, and then the whole value
+is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. This quoting logic will be present in the router.
+To quote using single-quotes, we will do the following:
+Replace ' with '\''.
+Wrap the entire string in single quotes.
+eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''
+
+
+### Risks and Mitigations
+
+We identified the following headers that require special explanations:
+
+#### Proxy header
+  As per haproxy security vulnerability, if the client sets `proxy` header, then the cgi app takes the value and injects it in the env var `HTTP_PROXY`,
+  so basically client is setting the proxy which is a security vulnerability as it is modifies the proxy value of the
+  environment variable which may be used by other applications so that is why we deny setting/deleting proxy header with this EP.
+  Please note that proxy headers are deleted by haproxy template by default.
+  `proxy` isn't necessarily a security risk (route owners could inject configuration into their own applications) but is non-standard and potentially
+  confusing, so it should be blocked.
+
+#### Strict-Transport-Security
+Strict-transport-security is already handled separately and we don't want to
+allow overrides of that configuration. If we allowed it, then route owners
+could impact each other's routes. It must be blocked so we have put this
+header in the deny list.
+
+#### X-Forwarded-For
+  `x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
+   another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
+  but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
+  and possibly a warning on admission.
+  Eg: If someone creates a Ingress Controller with the following:
+```sh
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+name: default
+namespace: openshift-ingress-operator
+spec:
+httpHeaders:
+forwardedHeaderPolicy: Replace
+```
+  Then the set-header stanzas will be added in the haproxy.config using the [haproxy-config.template](https://github.com/openshift/router/edit/master/images/router/haproxy/conf/haproxy-config.template#L565-L580)
+  in the plain backend section of the haproxy.config.
+  However, if someone wants to set custom values for these then they need to use the API of this EP.
+  If both are set then custom values will takeover.
```

Comment:
The word "takeover" is a noun; use "take over" for the verb.  
```suggestion
  If both are set then custom values will take over.
```

Reply:
done